src/server/session.ts

```diff
@@ -120,6 +120,117 @@ namespace ts.server {
         export const Unknown = "unknown";
     }
 
+    namespace Metrics {
+        let eventCounts: Map<number> = {};
+        let properties: Map<string> = {};
+        let dtsCount = 0;
+
+        const sendInterval = 1000 * 60 * 5; // 5 minutes
+        let nextSendTimeMs = Date.now() + sendInterval;
+
+        const settingNames = [
+            "allowNonTsExtensions",
+            "declaration",
+            "emitBOM",
+            "emitDecoratorMetadata",
+            "experimentalAsyncFunctions",
+            "jsx",
+            "isolatedModules",
+            "module",
+            "moduleResolution",
+            "noEmit",
+            "noEmitOnError",
+            "noImplicitAny",
+            "noLib",
+            "noResolve",
+            "out",
+            "outFile",
+            "preserveConstEnums",
+            "removeComments",
+            "suppressImplicitAnyIndexErrors",
+            "target"
+        ];
+
+        // Keeps a count of total invocations of various language service operations (rename, gotodef, etc)
+        export function countEvent(eventName: string, projectSvc: ProjectService, host: ts.System) {
+            const opts = projectSvc.getFormatCodeOptions();
+            if (opts.SendMetrics) {
+                eventCounts[eventName] = (eventCounts[eventName] || 0) + 1;
+
+                if (Date.now() > nextSendTimeMs) {
+                    registerSettings(projectSvc);
+                    send(host, opts.TelemetryUserID);
+                }
+            }
+        }
+
+        // Keeps track of any .d.ts files used which correspond to known versions from DefinitelyTyped
+        export function countDts(file: string) {
+            for (const p in properties) {
+                if (properties[p] === file) {
+                    return;
+                }
+            }
+            dtsCount++;
+            properties["dtf" + dtsCount] = file;
+        }
+
+        function registerSettings(svc: ProjectService) {
+            properties["host"] = svc.hostConfiguration.hostInfo;
+            properties["inferredProjects"] = svc.inferredProjects.length.toString();

```

Comment:
what does this buy us? would it be better to send an event everytime we create project, instead of sending the count everytime?